### PR TITLE
Add support for manual OIDC authentication

### PR DESCRIPTION
**Pull request by pehala**
- Add support for manual OIDC authentication.
  - Meaning user will have to add bearer token into headers manually, sadly full OIDC workflow would introduce another dependency used just for this.
- Extract adding retries to sessions into a separate method.
- Reintroduce `backoff_factor` into default `Session` object.
- Add `update` method to `OIDCConfig`.
- Move `oidc` from `Service` to `Proxy`.

### CODE REVIEW

**Comment by pehala** [comment]
@mganisin @jsmadis @pestanko Could you please look at this?

**Comment by pehala** [comment]
- Add option to use custom authentication for application.
- Rebase to latest master.

@mganisin WDYT?

File: threescale_api/utils.py [hunk 1]
```diff
@@ -64,6 +61,14 @@ def __init__(self, app, endpoint: str = "sandbox_endpoint",
 
         logger.debug("[HTTP CLIENT] New instance: %s", self._baseurl)
 
+    @staticmethod
+    def retry_for_session(session: requests.Session, total: int = 4):
+        retry = Retry(total=total, backoff_factor=1, status_forcelist=(503,), raise_on_status=True,
```
Comment (mganisin) [inline threescale_api/utils.py]: raise_on_status should be rather False. this should be 'transparent' hence for the caller it should behave like it wasn't there (except of longer time due to retries)

File: threescale_api/utils.py [hunk 2]
```diff
@@ -49,12 +49,9 @@ def __init__(self, app, endpoint: str = "sandbox_endpoint",
...
         if session is None:
             session = requests.Session()
-            retry = Retry(total=4, status_forcelist=(503,), raise_on_status=False,
-                          respect_retry_after_header=False)
+            self.retry_for_session(session)
 
-            adapter = HTTPAdapter(max_retries=retry)
-            session.mount("https://", adapter)
-            session.mount("http://", adapter)
+        if app.authobj is not None:
```
Comment (mganisin) [inline threescale_api/utils.py]: this shouldn't be necessary anymore as we discussed that authobj can't be None unless it is correct value == authobj always returns right object or raises error

File: threescale_api/auth.py [hunk 3]
```diff
@@ -4,33 +4,35 @@
...
-        self.location = location
+        self.location = app.service.proxy.list().entity["credentials_location"]
 
     @property
     @abc.abstractmethod
     def credentials(self):
         "returns credentials pair as tuple"
 
-    def __call__(self, r):
+    def __call__(self, request):
```
Comment (mganisin) [inline threescale_api/auth.py]: this renaming doesn't seem to be relevant to the change in this PR, rather shouldn't be in this PR

File: threescale_api/resources.py [hunk 4]
```diff
@@ -417,7 +417,7 @@ def service(self) -> 'Service':
     def list(self, **kwargs):
         if "env" in kwargs:
             self._env = kwargs["env"]
-            del(kwargs["env"])
+            del (kwargs["env"])
```
Comment (mganisin) [inline threescale_api/resources.py]: this renaming doesn't seem to be relevant to the change in this PR, rather shouldn't be in this PR

File: threescale_api/resources.py [hunk 5]
```diff
@@ -490,6 +490,7 @@ def url(self) -> str:
     def update(self, params: dict = None, **kwargs) -> 'DefaultResource':
         return self.rest.patch(url=self.url, json=params, **kwargs)
 
+
```
Comment (mganisin) [inline threescale_api/resources.py]: this renaming doesn't seem to be relevant to the change in this PR, rather shouldn't be in this PR

File: threescale_api/resources.py [hunk 6]
```diff
@@ -726,6 +726,7 @@ def test_request(self, relpath=None, verify: bool = None):
 
         return client.get(relpath)
 
+
```
Comment (mganisin) [inline threescale_api/resources.py]: this renaming doesn't seem to be relevant to the change in this PR, rather shouldn't be in this PR

File: threescale_api/auth.py [hunk 7]
```diff
@@ -4,33 +4,35 @@
 
 import requests
 
+
```
Comment (mganisin) [inline threescale_api/auth.py]: this renaming doesn't seem to be relevant to the change in this PR, rather shouldn't be in this PR

File: threescale_api/auth.py [hunk 8]
```diff
@@ -4,33 +4,35 @@
...
 class BaseClientAuth(requests.auth.AuthBase):
     "Abstract class for authentication of api client"
     def __init__(self, app, location):
         self.app = app
-        self.location = location
+        self.location = app.service.proxy.list().entity["credentials_location"]
```
Comment (mganisin) [inline threescale_api/auth.py]: we need to consult this in person ;)

File: threescale_api/resources.py [hunk 9]
```diff
@@ -652,6 +656,10 @@ def __init__(self, entity_name='system_name', **kwargs):
 class Application(DefaultResource):
     def __init__(self, entity_name='system_name', **kwargs):
         super().__init__(entity_name=entity_name, **kwargs)
+        self.authobj_factories = {
```
Comment (mganisin) [inline threescale_api/resources.py]: we talked about underscore in python and agreed on its disputable purpose, anyway I would prefer this to be prefixed with _ to "hide" it (unless proven otherwise)
Also I see this isn't dict of factories anymore (thank you for this), maybe name could be adjusted (to have it more generic, 'typeless'?)

File: threescale_api/auth.py (hunk 8 above)
Comment (pehala) [reply threescale_api/auth.py]: I think this is resolved based on our conversation.

File: threescale_api/resources.py (hunk 9 above)
Comment (pehala) [reply threescale_api/resources.py]: Renamed

File: threescale_api/auth.py (hunk 7 above)
Comment (pehala) [reply threescale_api/auth.py]: Fixed

File: threescale_api/resources.py (hunk 6 above)
Comment (pehala) [reply threescale_api/resources.py]: Fixed

File: threescale_api/resources.py (hunk 4 above)
Comment (pehala) [reply threescale_api/resources.py]: Fixed

File: threescale_api/auth.py (hunk 3 above)
Comment (pehala) [reply threescale_api/auth.py]: Resolved based on our conversation

File: threescale_api/utils.py (hunk 2 above)
Comment (pehala) [reply threescale_api/utils.py]: Fixed

File: threescale_api/utils.py (hunk 1 above)
Comment (pehala) [reply threescale_api/utils.py]: Fixed

File: threescale_api/resources.py (hunk 5 above)
Comment (pehala) [reply threescale_api/resources.py]: Fixed

**Comment by pehala** [comment]
Applied fixed based on the feedback

@mganisin WDYT?
